test/common/upstream/conn_pool_map_impl_test.cc

```diff
@@ -0,0 +1,233 @@
+
+  Http::ConnectionPool::Instance& pool1 = test_map->getPool(1, getBasicFactory());
+  Http::ConnectionPool::Instance& pool2 = test_map->getPool(2, getBasicFactory());
+
+  EXPECT_EQ(&pool1, &test_map->getPool(1, getBasicFactory()));
+  EXPECT_EQ(&pool2, &test_map->getPool(2, getBasicFactory()));
+  EXPECT_EQ(test_map->size(), 2);
+}
+
+TEST_F(ConnPoolMapImplTest, TestEmptyClerWorks) {
+  TestMapPtr test_map = makeTestMap();
+
+  test_map->clear();
+  EXPECT_EQ(test_map->size(), 0);
+}
+
+TEST_F(ConnPoolMapImplTest, TestClearEmptiesOutMap) {
+  TestMapPtr test_map = makeTestMap();
+
+  test_map->getPool(1, getBasicFactory());
+  test_map->getPool(2, getBasicFactory());
+
+  test_map->clear();
+  EXPECT_EQ(test_map->size(), 0);
+}
+
+TEST_F(ConnPoolMapImplTest, CallbacksPassedToPools) {
+  TestMapPtr test_map = makeTestMap();
+
+  test_map->getPool(1, getBasicFactory());
+  test_map->getPool(2, getBasicFactory());
+  Http::ConnectionPool::Instance::DrainedCb cb1;
+  EXPECT_CALL(*mock_pools_[0], addDrainedCallback(_)).WillOnce(SaveArg<0>(&cb1));
+  Http::ConnectionPool::Instance::DrainedCb cb2;
+  EXPECT_CALL(*mock_pools_[1], addDrainedCallback(_)).WillOnce(SaveArg<0>(&cb2));
+
+  ReadyWatcher watcher;
+  test_map->addDrainedCallback([&watcher] { watcher.ready(); });
+
+  EXPECT_CALL(watcher, ready()).Times(2);
+  cb1();
+  cb2();
+}
+
+// Tests that if we add the callback first, it is passed along when pools are created later.
+TEST_F(ConnPoolMapImplTest, CallbacksCachedAndPassedOnCreation) {
+  TestMapPtr test_map = makeTestMap();
+
+  ReadyWatcher watcher;
+  test_map->addDrainedCallback([&watcher] { watcher.ready(); });
+
+  Http::ConnectionPool::Instance::DrainedCb cb1;
+  test_map->getPool(1, getFactoryExpectDrainedCb(&cb1));
+
+  Http::ConnectionPool::Instance::DrainedCb cb2;
+  test_map->getPool(2, getFactoryExpectDrainedCb(&cb2));
+
+  EXPECT_CALL(watcher, ready()).Times(2);
+  cb1();
+  cb2();
+}
+
+// Tests that if we drain connections on an empty map, nothing happens.
+TEST_F(ConnPoolMapImplTest, EmptyMapDrainConnectionsNop) {
+  TestMapPtr test_map = makeTestMap();
+  test_map->drainConnections();
+}
+
+// Tests that we forward drainConnections to the pools.
+TEST_F(ConnPoolMapImplTest, DrainConnectionsForwarded) {
+  TestMapPtr test_map = makeTestMap();
+
+  test_map->getPool(1, getBasicFactory());
+  test_map->getPool(2, getBasicFactory());
+  EXPECT_CALL(*mock_pools_[0], drainConnections());
+  EXPECT_CALL(*mock_pools_[1], drainConnections());
+
+  test_map->drainConnections();
+}
+
+TEST_F(ConnPoolMapImplTest, ClearDefersDelete) {
+  TestMapPtr test_map = makeTestMap();
+
+  test_map->getPool(1, getBasicFactory());
+  test_map->getPool(2, getBasicFactory());
+  test_map->clear();
+
+  EXPECT_EQ(dispatcher_.to_delete_.size(), 2);
+}
+
+// The following tests only die in debug builds, so don't run them if this isn't one.
+#if !defined(NDEBUG)
+TEST_F(ConnPoolMapImplTest, ReentryClearTripsAssert) {
```

Comment:
nit: Please name these DeathTest (you can subclass the main test just for the name, see other examples in the code base).